source/secure_tunneling_operations.c

```diff
@@ -0,0 +1,670 @@
+    storage_view->service_id = message_options->service_id;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->service_id)) {
+        return AWS_OP_ERR;
+    }
+
+    storage_view->payload = message_options->payload;
+    if (aws_byte_buf_append_and_update(&message_storage->storage, &storage_view->payload)) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_secure_tunnel_message_storage_clean_up(struct aws_secure_tunnel_message_storage *message_storage) {
+    aws_byte_buf_clean_up(&message_storage->storage);
+}
+
+/* Sets the stream id on outbound message based on the service id (or lack of for V1) to the current one being used. */
+static int s_aws_secure_tunnel_operation_message_set_stream_id(
+    struct aws_secure_tunnel_operation *operation,
+    struct aws_secure_tunnel *secure_tunnel) {
+
+    struct aws_secure_tunnel_operation_message *message_op = operation->impl;
+    int32_t stream_id = INVALID_STREAM_ID;
+
+    struct aws_secure_tunnel_message_view *message_view = &message_op->options_storage.storage_view;
+
+    if (message_view->service_id.len > 0) {
+        struct aws_string *service_id = NULL;
+        service_id = aws_string_new_from_cursor(secure_tunnel->allocator, &message_view->service_id);
+
+        if (secure_tunnel->config->service_id_1 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_1, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_1_stream_id;
+        } else if (
+            secure_tunnel->config->service_id_2 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_2, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_2_stream_id;
+        } else if (
+            secure_tunnel->config->service_id_3 != NULL &&
+            aws_string_compare(secure_tunnel->config->service_id_3, service_id) == 0) {
+            stream_id = secure_tunnel->config->service_id_3_stream_id;
+        } else {
+            /* service_id doesn't match any existing service id*/
+            AWS_LOGF_DEBUG(
+                AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+                "id=%p: invalid service_id:%s attempted to be used with an outbound message",
+                (void *)message_view,
+                aws_string_c_str(service_id));
+            stream_id = INVALID_STREAM_ID;
+        }
+        aws_string_destroy(service_id);
+    } else {
+        stream_id = secure_tunnel->config->stream_id;
+    }
+
+    if (stream_id == INVALID_STREAM_ID) {
+        return aws_raise_error(AWS_ERROR_IOTDEVICE_SECURE_TUNNELING_INVALID_STREAM);
+    }
+
+    message_op->options_storage.storage_view.stream_id = stream_id;
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * Check the outbound stream start service id (or lack of one for V1) and set the secure tunnel and stream start
+ * message's stream id to the next value.
+ */
+static int s_aws_secure_tunnel_operation_message_set_next_stream_id(
```

Comment:
Comments above apply to this function too.

Also, I just don't understand the semantics of this function.  It takes a service id (as specified in a message) and then uses a different, larger stream id (written back into that message) than what is assigned to that service id.  Larger context of why this is happening would be helpful.

Reply:
The Destination device connects, receives service ids (multiplexing) from the service, and each of the (up to 3) service id's stream ids are set to 0 (unused). 

When a Source device connects on the other end of the tunnel, it starts a stream by sending a service id along with whatever number they want as the stream id. 

When the function above (s_aws_secure_tunnel_operation_message_set_stream_id) receives a stream start message, it checks against the service ids, and sets that service id's stream id to the received one from the stream start. 

This function here (s_aws_secure_tunnel_operation_message_set_next_stream_id ) is only called from Source mode sending a Stream Start message and it gets the next stream id to use and sets a its own stream id to the same one so it can send messages and receive messages on the right stream id.